### ISSUE TITLE
attempted voting for pipeline proposal

### ISSUE THREAD

**Issue by aadamsx** [tc39/proposal-pipeline-operator#163]
Due to overwhelming commentary that voting is not the route to take to get consensus, I've taken down the poll.

**Comment by nafg** [comment]
I accidentally voted twice. I didn't realize you could embed a poll within a GitHub issue

**Comment by aikeru** [comment]
With all due respect - and I want to see some flavor of the pipeline advanced so much even if it's not my favorite flavor - TC39 is a committee, not a democracy. What meaning does this vote have to the committee? Has the committee endorsed holding a vote to decide which flavor of the pipeline operator will move forward?

If this is the path the committee would endorse, shouldn't this be widely publicized? Have you contacted newsletters, blogs, etc. to solicit more votes from the broader JavaScript community?

Also, I see the partial application proposal mentioned a lot with the F# style. I'd like to remind everyone that pushing the F# style through does not in any way mean getting partial application with it, and that partial application is still at stage 1 and has its own challenges moving forward.
Another point I'd like to make for the proponents wishing for this combination is that you can't use partial application's placeholder in an expression - it can only be used an an argument to a function. IE, this does not work...
```
5 |> setValue(5 + ?)
```
...this does...
```
5 |> setCoordinates(5, ?)
```

Finally, one last point I'd like to make - shouldn't we resolve the open issues with the F# proposal before trying to push it through?
For example, these issues are still open and seem to affect the F# flavor:
https://github.com/tc39/proposal-pipeline-operator/issues/104
https://github.com/tc39/proposal-pipeline-operator/issues/23
https://github.com/tc39/proposal-pipeline-operator/issues/82
https://github.com/tc39/proposal-pipeline-operator/issues/145
Whether the final proposal will require parenthesis wrapping arrow functions used with the pipeline operator is still an open issue ( https://github.com/tc39/proposal-pipeline-operator/issues/162#issuecomment-604716126 ).
Is the dangling await issue resolved? https://github.com/tc39/proposal-pipeline-operator/issues/83
Can we abandon inline arrows `->` for F#? https://github.com/tc39/proposal-pipeline-operator/issues/93
I'm not saying smart doesn't have issues, but the creator of this poll is clearly strongly in favor of F# and of dropping smart, which is why I bring this up.

**Comment by aadamsx** [comment]
With all due respect @aikeru you're jumping ahead here.

According to @ljharb (and others) here: https://github.com/tc39/proposal-pipeline-operator/issues/157#issuecomment-607946058

Paraphrasing:

> To move forward, this proposal we need to pick a solution shape “that the community prefers”

This vote is an attempt to get JS community feedback, and what better place to do that than the repository for this proposal.  Also, as I supsected, the F# style is preferred, and you can see it's not even close.

The next step is to take this consensus and got to the TC39 commity with our proposal, I'm working on this now.

(Also, everyone should know this is not about partial application, it's about pipes, so please don't muddy the waters.)

**Comment by aikeru** [comment]
@aadamsx you are actually misquoting that quote, and changing the meaning of it.

> To move forward, this proposal **would not just have to pick a solution shape “that the community prefers” - it also has to prepare an ironclad, persuasive argument** for why that shape is the only one that will make sense forever (including outlining potential extensions, or which extensions are precluded). 

Bold emphasis mine. And respectfully, while you and I want the same thing, no, I do not agree with this being the best place to both solicit that feedback and to publicize it. Look at the watch numbers and stars for this repository (224 and 4.6k respectively). This is not even a tiny fraction of the JavaScript community. If you want _community consensus_ that is meaningful, I think we are talking a large number of thousands.

You're basing your conclusion on what is, as of this writing, 20 votes for F#? That's not a community consensus. That's a single office in a small company.

Also, I'd like to see some additional questions on this - have the participants of this vote used any form of the proposal such as through Babel or in browsers that have implemented it? I'd like to at least see some basic comparisons between the two. Telling people "go read the issues" seems like you're shifting more effort to your audience at a time when you want them to fully participate and engage.

**Comment by aadamsx** [comment]
@aikeru I said I'm paraphrasing, which is NOT a quote guy.

**Comment by aikeru** [comment]
@aadamsx I wish to keep my responses civil and friendly. To that end, I wish you the best in your endeavor. I hope some flavor of this proposal wins sooner than later.

**Comment by aadamsx** [comment]
@aikeru you said I was misquoting, which is clearly not what I was doing (I said I was paraphrasing), so it was kinda not being civil to say I was doing something I wasn't.

**Comment by aadamsx** [comment]
@nafg
> I accidentally voted twice

You actually cannot vote twice (it might say thanks, but it doesn't increment).

**Comment by aadamsx** [comment]
@aikeru 

> shouldn't this be widely publicized? Have you contacted newsletters, blogs, etc. to solicit more votes from the broader JavaScript community?

> I do not agree with this being the best place to both solicit that feedback and to publicize it. 

Please, send me the links to the JS gathering places you'd like for me to solicit this vote?  @Anyone, please?  

I want to case as wide a net as possible!

**Comment by ljharb** [comment]
@aadamsx https://github.com/tc39/proposal-pipeline-operator/issues/163#issuecomment-608128321 is a correct interpretation of my comment. Community preference for a single solution is always a good thing to have, but is not sufficient. The much more important thing is the ironclad persuasive argument - focus on that before trying to gather aesthetic community preferences.

**Comment by aadamsx** [comment]
@ljharb

The way you phrased it here:

> To move forward, this proposal would not just have to pick a solution shape “that the community prefers” - it also has to prepare an ironclad, persuasive argument for why that shape is the only one that will make sense forever

Told me: 

1) proposal would not just have to pick a solution shape “that the community prefers”

The step I'm trying to accomplish now with this vote

2) it also has to prepare an ironclad, persuasive argument for why that shape is the only one that will make sense forever

Which, according to the voting, I'm going to starting working on, in the F# style.

I think my statement above makes this clear:

> The next step is to take this F# style pipes operator consensus and go to the TC39 committee with our proposal, I'm working on this now.

**Comment by ljharb** [comment]
@aadamsx i was replying to your implication that a selection was needed. If the reason for picking one of the options is "the community prefers it", that simply won't change anything. There have to be concrete reasons it's preferred, and a poll/survey isn't useful for that. TC39 does not make decisions based on surveys, polls, or votes.

**Comment by aadamsx** [comment]
> There have to be concrete reasons it's preferred, and a poll/survey isn't useful for that

No, but this is a way to get at what the "community prefers", which is the way it should be.  Presumably they prefer this way because of course they've researched the options and came to prefer one way over the other.

Now the next step is to solicit the "prepare an ironclad, persuasive argument", which is what I'm moving on now.  Does that make sense?

**Comment by aadamsx** [comment]
@ljharb 

> There have to be concrete reasons it's preferred

The problem is, there is nothing really concrete about this, it's all opinion and preference.  Both F# and Smart have their pros and cons, which is better depends on context, the style you prefer to write code in, etc.

How can the we move forward when the "right" one depends on who you're asking?  It seems the only/best way to reslove this is pitch all the ways this could be implemented to the "committee" and have them decide its fate?  

By the way are these panelists that decide the fate of JS these days? https://www.youtube.com/watch?v=slA06pbTRi4

**Comment by ljharb** [comment]
Keep in mind that “nothing” is a better choice than “the wrong thing forever”. If it’s truly that subjective, then we may want to not advance this proposal.

**Comment by aadamsx** [comment]
It's not the idea that's subjective @ljharb of course, it's the implementation.

**Comment by szTheory** [comment]
Just because there are multiple good ways to implement a feature would not be a good reason to block a proposal. It could just mean that you have to pick one. I would take smart pipelines over no pipelines.

**Comment by mikestopcontinues** [comment]
@aadamsx a few questions:

Is there anything that would prevent smart from being implemented on top of F#? 

If not, the case everyone can agree on is that we submit them as separate, paired proposals. This would also help, because the implementation of the smart placeholder (#) overlaps with the current placeholder proposal (?). So the second proposal would be free to focus on that issue. And the second question:

What's holding us back from making one or both proposals?

If it's time, I volunteer to do as much of the legwork/writing as I can. Pipelines have become invaluable to me, and I want to see them as part of the language. (With placeholders, if possible. :D )

**Comment by noppa** [comment]
@mikestopcontinues There's discussion about the differences between Smart and F# in #134  
AFAIK Smart can be implemented on top of F# if you make `x |> expressionWithoutPlaceholder` mean "pass x as the argument to the function that we get from evaluating expressionWithoutPlaceholder". It's currently a syntax error in Smart, because with those semantics the presence of `#` significantly changes how the RHS behaves and it could be hard to parse. Hard to parse for humans, that is, correct me if I'm wrong but I haven't seen anyone say that it'd be hard for a parser.

Personally, I'm skeptical that Smart would advance after F# had been accepted, given that there's a separate partial application proposal that's more general solution to the problem.

@ljharb 
I don't think anyone wants to rush half-assed solution out of the door. It's just that, from an outsider's perspective, issues like "community consensus" don't seem to be worked on by anyone and the only possible outcome of this proposal looks to be a permanent deadlock. Lots of people don't want to see that happen and are eager to help move things forward, but don't know how. It's not clear what it'd take to be able to say that "solution X has community consensus", which makes it neigh impossible to help ever reach that goal.

**Comment by Pokute** [comment]
> Keep in mind that “nothing” is a better choice than “the wrong thing forever”.

I believe that there's a sizeable portion of pipeline operator hopefuls who think that there is no _wrong_ choice between F# and smart pipelines for the end-user experience. From engine implementation / spec / compatibility perspective there might be such a wrong choice, but for those I expect that the proposal process will detect them first.

If a poll on pipeline implementation preference was conducted, I would like to see a mandatory second choice in the selections: smart preference over F#, F# preference over smart, F# or none, smart or bust, neither. If the count for first two selections have a majority then, in the case where both proposals end up with no winner otherwise, the final choice can be made fairly and justly with a coin toss.

I'm hoping that the proposal process will clean up both pipeline implementations to a point where after implementing one, regardless of whichever ends up being chosen, it is recognised as suboptimal forever at worst.

**Comment by pygy** [comment]
@ljharb why can't the minimal proposal move forward as is (spec'ed in a way that makes it forward compatible with either smart or F#)?

This approach is a figment of the Rust design process, and they use it with great success.

Also, the pipeline is useful for most users, but most critical for FP enthusiasts, and the minimal proposal would fit the common use case for that population.

**Comment by ljharb** [comment]
Establishing that ability to augment it later would help a lot - but there’s still folks who don’t think either one is worth the syntax cost - they have to be convinced, and only meeting a subset of the use cases by intentionally shipping a subset of the semantics doesn’t seem to me like it’d be compelling enough to convince them.

**Comment by pygy** [comment]
Who are those folks?

Generally, provided the TC-39 thought that adding classes would be a good idea, I knew that promoting this operator would be an uphill battle, but I didn't expect this to last this long.

Imagine classes being derailed by FP folks who think that the objects they create should be immutable. This is what this proposal feels like.

Also, the syntax cost is minimal, and, unlike classes (TC-39: "You won't have to use them if you don't like them". WhatWG: "Hold my beer..."), they are purely local. The rest of the world won't know you use it internally, you can't expose them as part of your API.

**Comment by yordis** [comment]
Related to what @phaux said: https://github.com/tc39/proposal-pipeline-operator/issues/167#issuecomment-746780798

Separate partial application placeholding out of the pipe operator would move the conversation forward. I would propose to keep partial application placeholder out of the spec and decide later about such topic.

**Comment by tabatkins** [comment]
Closing this issue, as the proposal has advanced to stage 2 with Hack-style syntax.

**Comment by aadamsx** [comment]
Yep, just as I told everyone, as soon as you took this project on, and I saw you were biased to hack style, fsharp style would fall. I won’t be using this feature now. Just hope you don’t get it past as hack style is garbage.
